Create GLTF component

snagy opened the pull request:
I'm guessing (from the glb in the repo) this is already in progress, but here's a first cut at a glTF component.  It supports gltf or GLB files.

It supports draco and KTX2 (basis) compression, but those have to be referenced at runtime and the loaders are in the static folder.  There is probably a better way to do this.

Animations are not supported yet.

Commit:
Create initial glTF component
* include Basis and Draco loaders
* Animations don't do anything yet.

DefinitelyMaybe commented:
Would like to have a further look into this but haven't been able to check in on the inspection from the bot. let us know if there's anything you'd like some help on

edit: Also a GLTF component sounds good to me 👍

snagy commented:
I get a 404 page when I try to check the Vercel error.  I'll look into it more later tonight.

aewing commented on src/lib/components/objects/GLTF.svelte:
```diff
@@ -0,0 +1,65 @@
+	const { root, parent, self } = setup(new THREE.Object3D());
+
+	$: {
+		if(gltfScene) {
```
Edit: Maybe worth dispatching an event on load/change as well.

snagy replied:
Can you give an example of what that might look like?

aewing replied:
Sure! 
At the top of the file before/after the props are exported, we could create an event dispatcher:
```js
import { createEventDispatcher } from "svelte";
const dispatch = createEventDispatcher();
```
And then on [line 77](https://github.com/Rich-Harris/svelte-cubed/pull/25/files#diff-881ff166de0eb6fb0745293f466a247ab3aa642fe5dab6f27960f794ebee8285R77) where we have successfully loaded (or failed to load) the model, we could `dispatch("load", gltf);` or `dispatch("error", err);`, etc...

My thinking was that this could provide a convenient entry point for manipulating the materials in the scene, etc...

aewing replied:
As a consumer of the component, that would look something like
```svelte
<SC.GLTF {...} on:load={swapMyTexture} on:error={swapFallbackScene} />
```

aewing replied:
I'm not sure that dispatching on **change** would provide any value to _me_, but the implementation would be similar, just within the reactive statement.

aewing commented on src/lib/components/objects/GLTF.svelte:
```diff
@@ -0,0 +1,65 @@
+
+	onMount(() => {
+
+		const loader = new GLTFLoader()
```
Is there a benefit to creating the loader for each component instance rather than reusing?

DefinitelyMaybe replied:
I hear what your saying but also just doing a thing and getting it off the ground/running is step # 1

aewing replied:
And reviewing it and iterating on it is step # 2!

snagy replied:
Absolutely!  I submitted this PR early to get feedback, so I'm excited to get feedback.  I looked into the loader and figured out how to share it; i'll push a version with this in a bit.  This also shares the KTX and Draco loaders.

DefinitelyMaybe replied:
ah! all good then. looking forward to previewing the deploy when that comes together

Commit:
Share loaders between GLTF instances.
Fix path so it can be left blank.
Added cesium man animated glb (not animated in component yet)

Commit:
Fix some linter stuff

snagy commented:
Can anyone see the Vercel deployment error?  i'm still getting a 404 when I try to see the details.    @Rich-Harris any ideas?

Myrmod commented:
For the time being I created a pull request with an example on how to load a gltf or glb file using `SC.Primitive` in #41.
